File: tests/C++/test_unimod.cpp
// [UNIMOD:<id>] modification-aware peptide sequence parsing (fasta_mods.h,
// unimod.h). See docs/ai/unimod.md for the notation and the shipped table's
// exclusions.

#include <cstdio>
#include <fstream>
#include <sstream>
#include <string>

#include "doctest.h"
#include "element_lookup.h"
#include "fasta.h"
#include "fasta_mods.h"
#include "isoSpec++.h"
#include "test_helpers.h"
#include "unimod.h"

using namespace IsoSpec;
using namespace test_helpers;

TEST_CASE("FromFASTAWithMods matches FromFASTA for a plain, unmodified sequence") {
    for (const char* seq : {"PEPTIDE", "MKWVTFISLLLLFSSAYSRGV", ""}) {
        INFO("sequence='" << seq << "'");
        Iso with_mods = Iso::FromFASTAWithMods(seq);
        Iso plain = Iso::FromFASTA(seq);
        CHECK(with_mods.getMonoisotopicPeakMass() == doctest::Approx(plain.getMonoisotopicPeakMass()));
        CHECK(with_mods.getTheoreticalAverageMass() == doctest::Approx(plain.getTheoreticalAverageMass()));

        Iso with_mods_dry = Iso::FromFASTAWithMods(seq, false, false);
        Iso plain_dry = Iso::FromFASTA(seq, false, false);
        CHECK(with_mods_dry.getMonoisotopicPeakMass() == doctest::Approx(plain_dry.getMonoisotopicPeakMass()));
    }
}

TEST_CASE("internal UNIMOD mod applies its composition delta") {
    // Carbamidomethyl, UNIMOD:4 -> H3C2N1O1, 57.021464 Da.
    Iso modified = Iso::FromFASTAWithMods("MPEPTC[UNIMOD:4]DEK");
    Iso base = Iso::FromFASTA("MPEPTCDEK");
    double delta = modified.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass();
    CHECK(delta == doctest::Approx(57.021464).epsilon(1e-5));

    // The exact regression this feature exists to fix: before, every letter
    // in "UNIMOD" (U, N, I, M, O, D) was misread as a phantom amino acid --
    // confirmed empirically against the unfixed code this session. The fix
    // must differ from the unmodified baseline by exactly Carbamidomethyl's
    // delta, not by six extra residues (which would be off by ~350+ Da).
    CHECK(delta < 100.0);
}

TEST_CASE("N-terminal UNIMOD mod applies before the first residue") {
    // Acetyl, UNIMOD:1 -> H2C2O1, 42.010565 Da.
    Iso modified = Iso::FromFASTAWithMods("[UNIMOD:1]-PEPTIDE");
    Iso base = Iso::FromFASTA("PEPTIDE");
    CHECK(modified.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass() ==
          doctest::Approx(42.010565).epsilon(1e-5));
}

TEST_CASE("C-terminal UNIMOD mod applies after the last residue") {
    // Oxidation, UNIMOD:35 -> O1, 15.994915 Da.
    Iso modified = Iso::FromFASTAWithMods("PEPTIDE-[UNIMOD:35]");
    Iso base = Iso::FromFASTA("PEPTIDE");
    CHECK(modified.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass() ==
          doctest::Approx(15.994915).epsilon(1e-5));
}

TEST_CASE("a real shipped composition mixing a 2-letter and adjacent 1-letter element parses correctly") {
    // UNIMOD:123 ICAT-H -> H20C15N1O6Cl1 -- exercises the tokenization
    // boundary a 2-letter symbol (Cl) next to a 1-letter one (C, N, O, H)
    // depends on: every element's count must be explicit (never bare "N"),
    // or "...O6Cl1" could misparse. Cross-checked against the hand-built
    // formula for the same composition.
    Iso modified = Iso::FromFASTAWithMods("PEPTC[UNIMOD:123]DEK");
    Iso base = Iso::FromFASTA("PEPTCDEK");
    Iso expected_delta("H20C15N1O6Cl1");
    double observed = modified.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass();
    // Iso(formula) adds no water of its own, so the modification's delta
    // compares directly against expected_delta's raw monoisotopic mass.
    CHECK(observed == doctest::Approx(expected_delta.getMonoisotopicPeakMass()).epsilon(1e-5));
}

TEST_CASE("a net-negative element count is returned by the parser, refused by Iso") {
    // Met->Hsl (UNIMOD:11) is H-4C-1S-1; a lone glycine has neither the
    // hydrogens nor the sulphur to pay for it. The parser's job is
    // arithmetic, so it reports the negative net (a further modification
    // could still bring it back up); the refusal belongs where a composition
    // becomes a molecule. Both halves are contract -- fasta_mods.h says so,
    // and the C ABI and R bindings rely on the first half being true.
    ElementComposition composition = parse_fasta_with_mods("G[UNIMOD:11]");
    bool saw_negative = false;
    for (size_t i = 0; i < composition.count.size(); i++)
        if (composition.count[i] < 0)
            saw_negative = true;
    CHECK(saw_negative);

    CHECK_THROWS_AS(build_iso_from_composition(composition), std::invalid_argument);
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("G[UNIMOD:11]", false, false), std::invalid_argument);
}

TEST_CASE("unknown UNIMOD id throws") {
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:999999999]DEK"), std::invalid_argument);
}

TEST_CASE("deliberately excluded UNIMOD id (isotope-labeled) throws, same as unknown") {
    // UNIMOD:9, ICAT-G:2H(8) -- isotope-labeled, excluded from the shipped
    // table by scripts/build_unimod_table.py (see docs/ai/unimod.md).
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:9]DEK"), std::invalid_argument);
}

TEST_CASE("an id past 32 bits is unknown, not silently truncated") {
    // The id is parsed as a 64-bit value but UnimodTable::lookup takes an
    // unsigned int, so without an explicit range check 2^32 + 4 would wrap to
    // 4 and quietly apply Carbamidomethyl -- a wrong answer with no
    // diagnostic, which is the worst possible outcome for a parser whose
    // whole job is resolving ids.
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:4294967300]DEK"), std::invalid_argument);
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:8589934596]DEK"), std::invalid_argument);

    // And the id it would have wrapped onto still resolves normally.
    Iso modified = Iso::FromFASTAWithMods("PEPTC[UNIMOD:4]DEK");
    Iso base = Iso::FromFASTA("PEPTCDEK");
    CHECK(modified.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass() ==
          doctest::Approx(57.021464).epsilon(1e-5));
}

TEST_CASE("an implausibly large id in a CSV is rejected, not allocated") {
    // entries_ is indexed straight by id, so an unchecked id here is a
    // request for id * sizeof(UnimodEntry) bytes.
    CHECK_THROWS_AS(parse_unimod_csv("id,name,mono_mass,composition\n"
                                     "4000000000,Typo,1.0,H1\n"),
                    std::invalid_argument);
    // The largest id the shipped table actually uses still loads.
    CHECK(embedded_unimod_table().lookup(2147) != nullptr);
}

TEST_CASE("malformed UNIMOD brackets throw") {
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:4DEK"), std::invalid_argument);         // missing ']'
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMOD:]DEK"), std::invalid_argument);          // missing id
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[FOO:4]DEK"), std::invalid_argument);            // wrong prefix
    CHECK_THROWS_AS(Iso::FromFASTAWithMods("PEPTC[UNIMODX4]DEK"), std::invalid_argument);         // no colon
}

TEST_CASE("non-bracket characters keep the old lenient (silently-ignored) behavior") {
    // The fix targets '[' specifically -- everything parse_fasta already
    // tolerated (spacers, whitespace, indeterminate-formula codes) must keep
    // working exactly as before, including the documented "AE-DA" example.
    Iso base = Iso::FromFASTAWithMods("AEDA");
    for (const char* variant : {"AE-DA", "EAXXDA*", "AE DA", "ae\tda", "A E D A\n", "AEDA?!"}) {
        INFO("variant='" << variant << "'");
        Iso v = Iso::FromFASTAWithMods(variant);
        CHECK(v.getMonoisotopicPeakMass() == doctest::Approx(base.getMonoisotopicPeakMass()));
    }
}

TEST_CASE("unimod_db_path override replaces the embedded table") {
    const char* tmp_path = "build/test_unimod_override.csv";
    {
        std::ofstream f(tmp_path);
        f << "id,name,mono_mass,composition\n";
        f << "4,TestOverride,1.0,H1\n";  // deliberately NOT Carbamidomethyl's real composition
    }

    Iso overridden = Iso::FromFASTAWithMods("PEPTC[UNIMOD:4]DEK", false, true, tmp_path);
    Iso base = Iso::FromFASTA("PEPTCDEK");
    double delta = overridden.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass();
    Iso expected_delta("H1");
    CHECK(delta == doctest::Approx(expected_delta.getMonoisotopicPeakMass()).epsilon(1e-5));

    // And the embedded default is untouched by having loaded an override.
    Iso still_default = Iso::FromFASTAWithMods("PEPTC[UNIMOD:4]DEK");
    double default_delta = still_default.getMonoisotopicPeakMass() - base.getMonoisotopicPeakMass();
    CHECK(default_delta == doctest::Approx(57.021464).epsilon(1e-5));

    std::remove(tmp_path);
}

TEST_CASE("unimod_table_for_path caches by path") {
    const char* tmp_path = "build/test_unimod_cache.csv";
    {
        std::ofstream f(tmp_path);
        f << "id,name,mono_mass,composition\n";
        f << "1,X,1.0,H1\n";
    }

    // A named std::string (rather than passing tmp_path inline) avoids a
    // spurious -Wdangling-reference: the returned reference is to a
    // function-static, not to anything owned by the argument, but the
    // compiler's heuristic can't see that through an implicit conversion in
    // the same expression as the reference binding.
    const std::string path(tmp_path);
    const UnimodTable& first = unimod_table_for_path(path);
    const UnimodTable& second = unimod_table_for_path(path);
    CHECK(&first == &second);

    std::remove(tmp_path);
}

TEST_CASE("embedded table agrees with data/unimod.csv on disk") {
    std::ifstream f("../../data/unimod.csv");
    REQUIRE(f.good());
    std::stringstream buffer;
    buffer << f.rdbuf();
    UnimodTable from_disk = parse_unimod_csv(buffer.str());

    for (unsigned int id : {1u, 4u, 7u, 21u, 34u, 35u, 121u, 123u}) {
        INFO("id=" << id);
        const UnimodEntry* embedded_entry = embedded_unimod_table().lookup(id);
        const UnimodEntry* disk_entry = from_disk.lookup(id);
        REQUIRE(embedded_entry != nullptr);
        REQUIRE(disk_entry != nullptr);
        CHECK(embedded_entry->name == disk_entry->name);
        CHECK(embedded_entry->mono_mass == doctest::Approx(disk_entry->mono_mass));
        CHECK(embedded_entry->element_first_index == disk_entry->element_first_index);
        CHECK(embedded_entry->element_delta_count == disk_entry->element_delta_count);
    }
}

File: tests/Python/test_unimod.py
"""[UNIMOD:<id>] modification-aware peptide sequence parsing.

ParsePeptideSequence / ParseFASTA (the compatibility alias) / Iso's
peptide_sequence=/fasta= kwargs -- see docs/ai/unimod.md for the notation,
the shipped table's exclusions, and the fasta/peptide_sequence naming note.
"""

import math
import os

import pytest

import IsoSpecPy
from IsoSpecPy.IsoSpecPy import ParseFASTA, ParsePeptideSequence


def test_parse_peptide_sequence_matches_parse_fasta_for_plain_sequence():
    assert ParsePeptideSequence("PEPTIDE") == ParseFASTA("PEPTIDE")


def test_peptide_sequence_kwarg_is_alias_for_fasta_kwarg():
    from_fasta = IsoSpecPy.Iso(fasta="PEPTIDE")
    from_peptide_sequence = IsoSpecPy.Iso(peptide_sequence="PEPTIDE")
    assert math.isclose(from_fasta.getMonoisotopicPeakMass(),
                        from_peptide_sequence.getMonoisotopicPeakMass())


def test_fasta_and_peptide_sequence_kwargs_conflict_raises():
    with pytest.raises(ValueError):
        IsoSpecPy.Iso(fasta="PEPTIDE", peptide_sequence="OTHERSEQ")


def test_unimod_bracket_regression_reproduces_this_sessions_manual_check():
    # Before this feature, every letter in "UNIMOD" (U, N, I, M, O, D) was a
    # valid 1-letter amino-acid code, so a [UNIMOD:<id>] bracket was silently
    # misread as six phantom residues instead of a modification. Confirmed
    # empirically against the unfixed code this session:
    #   ParseFASTA('MPEPTC[UNIMOD:4]DEK') -> {'C': 76, 'H': 123, 'N': 19,
    #                                          'O': 27, 'S': 3, 'Se': 1}
    # (wrong) instead of the base composition plus Carbamidomethyl's delta.
    bracketed = ParseFASTA("MPEPTC[UNIMOD:4]DEK")
    base = ParseFASTA("MPEPTCDEK")

    assert "Se" not in bracketed  # the old bug's tell: phantom Se from misread "U"

    carbamidomethyl_delta = {"H": 3, "C": 2, "N": 1, "O": 1}
    expected = dict(base)
    for symbol, delta in carbamidomethyl_delta.items():
        expected[symbol] = expected.get(symbol, 0) + delta
    assert dict(bracketed) == expected


def test_formula_and_peptide_sequence_with_mods_combine():
    both = IsoSpecPy.Iso(formula="H2O1", peptide_sequence="PEPTC[UNIMOD:4]DEK")
    seq_only = IsoSpecPy.Iso(peptide_sequence="PEPTC[UNIMOD:4]DEK")
    water = IsoSpecPy.Iso("H2O1")
    assert math.isclose(
        both.getMonoisotopicPeakMass(),
        seq_only.getMonoisotopicPeakMass() + water.getMonoisotopicPeakMass())


def test_unknown_unimod_id_raises():
    with pytest.raises(ValueError):
        ParsePeptideSequence("PEPTC[UNIMOD:999999999]DEK")


def test_excluded_unimod_id_raises_same_as_unknown():
    # UNIMOD:9, ICAT-G:2H(8) -- isotope-labeled, deliberately excluded from
    # the shipped table (see scripts/build_unimod_table.py).
    with pytest.raises(ValueError):
        ParsePeptideSequence("PEPTC[UNIMOD:9]DEK")


def test_malformed_bracket_raises():
    with pytest.raises(ValueError):
        ParsePeptideSequence("PEPTC[UNIMOD:4DEK")


def test_unimod_db_path_override(tmp_path):
    override = tmp_path / "custom_unimod.csv"
    override.write_text("id,name,mono_mass,composition\n4,TestOverride,1.0,H1\n")

    overridden = ParsePeptideSequence("PEPTC[UNIMOD:4]DEK", unimod_db_path=str(override))
    base = ParseFASTA("PEPTCDEK")
    expected = dict(base)
    expected["H"] = expected.get("H", 0) + 1
    assert dict(overridden) == expected

    # The embedded default is untouched by having loaded an override.
    still_default = ParsePeptideSequence("PEPTC[UNIMOD:4]DEK")
    assert still_default["C"] - base["C"] == 2  # Carbamidomethyl's real C delta, not the override's 0


def test_id_past_32_bits_is_unknown_not_truncated():
    # The id is parsed as 64-bit but looked up as unsigned int; without a
    # range check 2**32 + 4 wraps to 4 and silently applies Carbamidomethyl.
    for bad_id in (2**32 + 4, 2**33 + 4):
        with pytest.raises(ValueError):
            ParsePeptideSequence("PEPTC[UNIMOD:{}]DEK".format(bad_id))
    # ...and the id it would have wrapped onto is unaffected.
    assert ParsePeptideSequence("PEPTC[UNIMOD:4]DEK")["C"] - ParseFASTA("PEPTCDEK")["C"] == 2


def test_zero_count_elements_are_omitted():
    # Documented behaviour change from the pre-Unimod ParseFASTA, which
    # always emitted C/H/N/O/S keys: only non-zero counts appear now.
    composition = ParseFASTA("A")
    assert "S" not in composition
    assert composition["C"] == 3


def test_negative_net_count_is_reported_by_the_parser_and_refused_by_Iso():
    # Met->Hsl (UNIMOD:11) removes H4 C1 S1; a lone glycine cannot pay for it.
    composition = ParsePeptideSequence("G[UNIMOD:11]")
    assert composition["H"] < 0
    with pytest.raises(Exception):
        IsoSpecPy.Iso(peptide_sequence="G[UNIMOD:11]")


def test_charge_keeps_its_positional_slot():
    # peptide_sequence/unimod_db_path were added after charge precisely so
    # this keeps working: formula, get_confs, atomCounts, isotopeMasses,
    # isotopeProbabilities, use_nominal_masses, fasta, charge.
    positional = IsoSpecPy.Iso("H2O1", False, None, None, None, False, "", 2.0)
    keyword = IsoSpecPy.Iso(formula="H2O1", charge=2.0)
    assert math.isclose(positional.getMonoisotopicPeakMass(),
                        keyword.getMonoisotopicPeakMass())
    assert math.isclose(keyword.getMonoisotopicPeakMass(),
                        IsoSpecPy.Iso(formula="H2O1").getMonoisotopicPeakMass() / 2.0)
